src/SpendPermissionManager.sol

```diff
@@ -0,0 +1,318 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {IERC1271} from "openzeppelin-contracts/contracts/interfaces/IERC1271.sol";
+import {IERC20} from "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import {CoinbaseSmartWallet} from "smart-wallet/CoinbaseSmartWallet.sol";
+import {EIP712} from "solady/utils/EIP712.sol";
+
+/// @title SpendPermissionManager
+///
+/// @notice Allow spending native and ERC20 tokens with a spend permission.
+///
+/// @dev Allowance and spend values capped at uint160 ~ 1e48.
+///
+/// @author Coinbase (https://github.com/coinbase/smart-wallet-permissions)
+contract SpendPermissionManager is EIP712 {
+    /// @notice A spend permission for an external spender to spend an account's tokens.
+    struct SpendPermission {
+        /// @dev Smart account this spend permission is valid for.
+        address account;
+        /// @dev Entity that can spend user funds.
+        address spender;
+        /// @dev Token address (ERC-7528 ether address or ERC-20 contract).
+        address token;
+        /// @dev Timestamp this spend permission is valid after (unix seconds).
+        uint48 start;
+        /// @dev Timestamp this spend permission is valid until (unix seconds).
+        uint48 end;
+        /// @dev Time duration for resetting used allowance on a recurring basis (seconds).
+        uint48 period;
+        /// @dev Maximum allowed value to spend within a recurring period.
+        uint160 allowance;
+    }
+
+    /// @notice Period parameters and spend usage.
+    struct PeriodUsage {
```

Comment:
what do you think of `PeriodSpend` vs. `PeriodUsage`?

Reply:
like it